File: greedy_algorithms/greedy_algorithms.go
package main

import "fmt"

func main() {
	var stations = make(map[string][]string)
	stations["kone"] = []string{"id", "nv", "ut"}
	stations["ktwo"] = []string{"wa", "id", "mt"}
	stations["kthree"] = []string{"or", "nv", "ca"}
	stations["kfour"] = []string{"nv", "ut"}
	stations["kfive"] = []string{"ca", "az"}

	var states_needed = []string{"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}
	final_stations := covering(states_needed, stations)
	fmt.Println(final_stations)

}

func covering(sn []string, stations map[string][]string) (final_stations []string) {


	for len(sn) > 0 {
		var best_station string
		var states_covered []string
		for k := range stations {
			covered := and(sn, stations[k])
			if len(covered) > len(states_covered) {
				states_covered = covered
				best_station = k
			}

		}
		if best_station != "" {
			final_stations = append(final_stations, best_station)
			sn = removeData(sn, states_covered)
		}
	}
	return 

}

func and(s1, s2 []string) (res []string) {
	for _, i := range s1 {
		for _, j := range s2 {
			if i == j {
				res = append(res, i)
			}
		}
	}
	return res
}

func removeData(states_needed, covered []string) (res []string) {
	for _, data := range covered {
		states_needed = remove(states_needed, data)
	}
	return states_needed
}

func remove(states_needed []string, value string) (res []string) {
	for i, item := range states_needed {
		if item == value {
			states_needed = append(states_needed[:i], states_needed[i+1:]...)

		}
	}
	return states_needed
}

File: greedy_algorithms/greedy_algorithms.py

stations = {}
stations["kone"] = {"id", "nv", "ut"}
stations["ktwo"] = {"wa", "id", "mt"}
stations["kthree"] = {"or", "nv", "ca"}
stations["kfour"] = {"nv", "ut"}
stations["kfive"] = {"ca", "az"}

states_needed = {"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}

def covering(states_needed, stations):
    final_stations = set()
    while states_needed:
        best_station = None
        states_covered = {}
        for station, states_for_station in stations.items():
            covered = states_needed & states_for_station

            if len(covered) > len(states_covered) and station not in final_stations:
                best_station = station
                states_covered = covered
        if best_station is not None:
            states_needed -= states_covered
            final_stations.add(best_station)
            stations.pop(best_station)
        else:
            return None
    return final_stations


def test(states_needed, stations):
    final_result = set()
    while states_needed:
        best_station = None
        states_covered = {}
        for station, states_for_station in stations.items():
            covered = states_for_station & states_needed
            if station not in final_result and len(covered) > len(states_covered):
                states_covered = covered
                best_station = station
        if best_station is not None:
            final_result.add(best_station)
            stations.pop(best_station)
            states_needed -= states_covered
    return final_result


                
        




print(test(states_needed, stations))
# print(covering(states_needed, stations))
  
    